### PR TITLE
Breakout parameters for x.509 certificate login

### PR DESCRIPTION
**Pull request by nicholasjackson**
This pull requests breaks out the the allowed_names parameter for restricting a certificate into three separate parameters:

* allowed_common_names
* allowed_dns
* allowed_emails

The existing `allowed_names` parameter still exists however it has also been marked as deprecated in both the documentation and CLI.  @jefferai I made an assumption on this if you do not intend to eventually remove the `allowed_names` I can updated the docs.

This pull request also includes my previous commit to add URI SAN validation, before starting this I rebased master so everything should be up to date.

### CODE REVIEW

**Comment by jefferai** [comment]
Can you change `allowed_dns` to `allowed_dns_names` (or `allowed_dns_sans` or something similar)? `allowed_dns` could be interpreted to mean "allowed DNs", as in, the full unique IDs of a given cert.

**Comment by nicholasjackson** [comment]
Hey @jefferai I think allowed_dns_sans is more descriptive, would it make sense to name all these with the same convention so:
* allowed_dns_sans
* allowed_uri_sans
* allowed_email_sans

**Comment by nicholasjackson** [comment]
@jefferai sorry for the delay on the updates to this, I have been traveling and then managed to catch flu.  The latest commit updates things to clearly define the additional parameters as being part of the SAN extension as discussed in previous comments.

File: builtin/credential/cert/path_login.go [hunk 1]
```diff
@@ -324,7 +422,7 @@ func (b *backend) loadTrustedCerts(ctx context.Context, storage logical.Storage,
 	trustedNonCAs = make([]*ParsedCert, 0)
 	names, err := storage.List(ctx, "cert/")
 	if err != nil {
-		b.Logger().Error("failed to list trusted certs", "error", err)
+		b.Logger().Error("cert: failed to list trusted certs", "error", err)
```
[the remark this answers is not in the document]
Comment (jefferai) [reply builtin/credential/cert/path_login.go]: Same with the other logging statements below.

[the remark this answers is not in the document]
Comment (nicholasjackson) [reply builtin/credential/cert/path_login.go]: @jefferai I have no idea on that, I wonder if I had an old branch, pretty sure I just extended my previous pull request, might not have had changes from master applied, let me double check the Sha's

**Comment by jefferai** [comment]
@nicholasjackson I removed the old stuff that had crept in. Something I can't do easily since this is on your fork instead of the main repo: Can you `s/Sans/SANs`?

**Comment by nicholasjackson** [comment]
@jefferai 100%, I will do this first thing tomorrow morning, so this will be ready for you tomorrow

**Comment by jefferai** [comment]
Looks great! Thank you!

**Pull request merged**